File: src/test_core/fmg/core/mosaic/MosaicViewTest.java
package fmg.core.mosaic;

import java.util.Collection;

import org.junit.*;

import fmg.common.LoggerSimple;
import fmg.common.geom.SizeDouble;
import fmg.common.notifier.PropertyChangeExecutor;
import fmg.core.img.IImageView;
import fmg.core.mosaic.cells.BaseCell;
import io.reactivex.Flowable;

class MosaicTestView extends MosaicView<DummyImage, DummyImage, MosaicTestModel> {
    MosaicTestView() { super(new MosaicTestModel()); }
    @Override protected DummyImage createImage() { return new DummyImage(); }
    private int drawCount;
    int getDrawCount() { return drawCount; }
    @Override protected void drawModified(Collection<BaseCell> modifiedCells) {
        LoggerSimple.put("MosaicTestView::drawModified");
        ++drawCount;
    }
    @Override public void close() {
        super.close();
        getModel().close();
    }
}

public class MosaicViewTest {

    static final int TEST_SIZE_W = MosaicModelTest.TEST_SIZE_W;
    static final int TEST_SIZE_H = MosaicModelTest.TEST_SIZE_H;

    @BeforeClass
    public static void setup() {
        LoggerSimple.put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        LoggerSimple.put("> MosaicViewTest::setup");

        MosaicModelTest.ProjSettings();

        Flowable.just("UI factory inited...").subscribe(LoggerSimple::put);
    }

    @Before
    public void before() {
        LoggerSimple.put("======================================================");
    }
    @AfterClass
    public static void after() {
        LoggerSimple.put("======================================================");
        LoggerSimple.put("< MosaicViewTest closed");
        LoggerSimple.put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }

    @Test
    public void propertyChangedTest() {
        LoggerSimple.put("> MosaicTestView::propertyChangedTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(100, 1000,
            view -> {
                view.getModel().setSize(new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));
            }, (view, modifiedProperties) -> {
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_MODEL));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_MODEL).first);
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_SIZE ));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_SIZE ).first);
                Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_IMAGE));
                Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_IMAGE).first);
                Assert.assertEquals(                3 , modifiedProperties.size());
            });
    }

    @Test
    public void readinessAtTheStartTest() {
        LoggerSimple.put("> MosaicTestView::readinessAtTheStartTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(10, 1000,
            view -> {
                Assert.assertEquals(0, view.getDrawCount());
                Assert.assertNotNull(view.getImage());
                Assert.assertEquals(1, view.getDrawCount());
            }, (view, modifiedProperties) -> { });
    }

    @Test
    public void multipleChangeModelOneDrawViewTest() throws InterruptedException {
        LoggerSimple.put("> MosaicTestView::multipleChangeModelOneDrawViewTest");

        DummyImage[] img = { null };
        MosaicTestView[] v = { null };
        new PropertyChangeExecutor<>(() -> v[0] = new MosaicTestView(), false).run(100, 1000,
           view -> {
               Assert.assertEquals(0, view.getDrawCount());
               MosaicModelTest.changeModel(view.getModel());
           }, (view, modifiedProperties) -> {
               img[0] = view.getImage();
               Assert.assertNotNull(img[0]);
               Assert.assertEquals(1, view.getDrawCount());
           });

        // test no change
        new PropertyChangeExecutor<>(() -> v[0], false).run(100, 1000,
           view -> {
               view.setSize(new SizeDouble(TEST_SIZE_W, TEST_SIZE_H));
           }, (view, modifiedProperties) -> {
               Assert.assertEquals(img[0], view.getImage());
               Assert.assertEquals(1, view.getDrawCount());
           });

        // test change
        new PropertyChangeExecutor<>(() -> v[0]).run(100, 1000,
           view -> {
               view.setSize(new SizeDouble(TEST_SIZE_W + 1, TEST_SIZE_H));
           }, (view, modifiedProperties) -> {
               Assert.assertNotEquals(img[0], view.getImage());
               Assert.assertNotNull(view.getImage());
               Assert.assertEquals(2, view.getDrawCount());
           });
    }

    @Test
    public void oneNotificationOfImageChangedTest() {
        LoggerSimple.put("> MosaicTestView::oneNotificationOfImageChangedTest");

        new PropertyChangeExecutor<>(MosaicTestView::new).run(100, 1000,
           view -> {
               MosaicModelTest.changeModel(view.getModel());
           }, (view, modifiedProperties) -> {
               Assert.assertTrue  (                    modifiedProperties.containsKey(IImageView.PROPERTY_IMAGE));
               Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageView.PROPERTY_IMAGE).first);
               Assert.assertEquals(0, view.getDrawCount());
               view.getImage(); // call the implicit draw method
               Assert.assertEquals(1, view.getDrawCount());
           });
    }

}

File: src/test_core/fmg/core/mosaic/MosaicViewTest.cs
using System.Threading.Tasks;
using System.Collections.Generic;
using fmg.common;
using fmg.common.geom;
using fmg.common.notifier;
using fmg.core.mosaic.cells;
using DummyImage = System.Object;
using MosaicTestModel = fmg.core.mosaic.MosaicDrawModel<object>;

namespace fmg.core.mosaic {

    class MosaicTestView : MosaicView<DummyImage, DummyImage, MosaicTestModel> {
        internal MosaicTestView() : base(new MosaicTestModel()) { }
        protected override DummyImage CreateImage() { return new DummyImage(); }
        internal int DrawCount { get; private set; }
        protected override void DrawModified(ICollection<BaseCell> modifiedCells) {
            LoggerSimple.Put(nameof(MosaicTestView) + "::DrawModified");
            ++DrawCount;
        }
        protected override void Disposing() {
            base.Disposing();
            Model.Dispose();
        }
    }

    public abstract class MosaicViewTest {

        internal const int TEST_SIZE_W = MosaicModelTest.TEST_SIZE_W;
        internal const int TEST_SIZE_H = MosaicModelTest.TEST_SIZE_H;

        protected abstract void AssertEqual(int    expected, int    actual);
        protected abstract void AssertEqual(double expected, double actual, double delta);
        protected abstract void AssertEqual(object expected, object actual);
        protected abstract void AssertNotNull(object anObject);
        protected abstract void AssertTrue(bool condition);
        protected abstract void AssertNotEqual(object expected, object actual);

        public virtual void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put("> " + nameof(MosaicViewTest) + "::" + nameof(Setup));
        }

        public virtual void Before() {
            LoggerSimple.Put("======================================================");
        }

        public virtual void After() {
            LoggerSimple.Put("======================================================");
            LoggerSimple.Put("< " + nameof(MosaicViewTest) + " closed");
            LoggerSimple.Put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        public virtual async Task PropertyChangedTest() {
            LoggerSimple.Put("> " + nameof(MosaicViewTest) + "::" + nameof(PropertyChangedTest));

            await new PropertyChangeExecutor<MosaicTestView>(() => new MosaicTestView()).Run(100, 1000,
                view => {
                    view.Model.Size = new SizeDouble(TEST_SIZE_W, TEST_SIZE_H);
                }, (view, modifiedProperties) => {
                    AssertTrue (   modifiedProperties.ContainsKey(nameof(view.Model)));
                    AssertEqual(1, modifiedProperties[            nameof(view.Model)]);
                    AssertTrue (   modifiedProperties.ContainsKey(nameof(view.Size)));
                    AssertEqual(1, modifiedProperties[            nameof(view.Size)]);
                    AssertTrue (   modifiedProperties.ContainsKey(nameof(view.Image)));
                    AssertEqual(1, modifiedProperties[            nameof(view.Image)]);
                    AssertEqual(3, modifiedProperties.Count);
                });
        }

        public virtual async Task ReadinessAtTheStartTest() {
            LoggerSimple.Put("> " + nameof(MosaicViewTest) + "::" + nameof(ReadinessAtTheStartTest));

            await new PropertyChangeExecutor<MosaicTestView>(() => new MosaicTestView()).Run(10, 1000,
                view => {
                    AssertEqual(0, view.DrawCount);
                    AssertNotNull(view.Image);
                    AssertEqual(1, view.DrawCount);
                }, (view, modifiedProperties) => { });
        }

        public virtual async Task MultipleChangeModelOneDrawViewTest() {
            LoggerSimple.Put("> " + nameof(MosaicViewTest) + "::" + nameof(MultipleChangeModelOneDrawViewTest));

            DummyImage img = null;
            MosaicTestView v = null;
            await new PropertyChangeExecutor<MosaicTestView>(() => v = new MosaicTestView(), false).Run(100, 1000,
                view => {
                    AssertEqual(0, view.DrawCount);
                    MosaicModelTest.ChangeModel(view.Model);
                }, (view, modifiedProperties) => {
                    img = view.Image;
                    AssertNotNull(img);
                    AssertEqual(1, view.DrawCount);
                });

            // test no change
            await new PropertyChangeExecutor<MosaicTestView>(() => v, false).Run(100, 1000,
                view => {
                    view.Size = new SizeDouble(TEST_SIZE_W, TEST_SIZE_H);
                }, (view, modifiedProperties) => {
                    AssertEqual(true, ReferenceEquals(img, view.Image));
                    AssertEqual(1, view.DrawCount);
                });

            // test change
            await new PropertyChangeExecutor<MosaicTestView>(() => v).Run(100, 1000,
                view => {
                    v.Size = new SizeDouble(TEST_SIZE_W + 1, TEST_SIZE_H);
                }, (view, modifiedProperties) => {
                    AssertNotEqual(img, view.Image);
                    AssertNotNull(view.Image);
                    AssertEqual(2, view.DrawCount);
                });
        }

        public virtual async Task OneNotificationOfImageChangedTest() {
            LoggerSimple.Put("> " + nameof(MosaicViewTest) + "::" + nameof(OneNotificationOfImageChangedTest));

            await new PropertyChangeExecutor<MosaicTestView>(() => new MosaicTestView()).Run(300, 1000,
                view => {
                    MosaicModelTest.ChangeModel(view.Model);
                }, (view, modifiedProperties) => {
                    AssertTrue(     modifiedProperties.ContainsKey(nameof(view.Image)));
                    AssertEqual(1, modifiedProperties[            nameof(view.Image)]);
                    AssertEqual(0, view.DrawCount);
                    var img = view.Image; // call the implicit draw method
                    AssertEqual(1, view.DrawCount);
                });
        }

    }

}
